src/Integration.Vsix/SonarLintTracker/IssueTracker.cs

```diff
@@ -0,0 +1,207 @@
+
+            this.Initialize();
+        }
+
+        private void FileActionOccurred(object sender, TextDocumentFileActionEventArgs e)
+        {
+            if (e.FileActionType == FileActionTypes.DocumentRenamed)
+            {
+                var oldPath = FilePath;
+                FilePath = e.FilePath;
+                provider.Rename(oldPath, FilePath);
+            }
+            else if (e.FileActionType == FileActionTypes.ContentSavedToDisk)
+            {
+                provider.RequestAnalysis(FilePath, Charset);
+            }
+        }
+
+        private void Initialize()
+        {
+            textBuffer.ChangedLowPriority += this.OnBufferChange;
+            provider.AddIssueTracker(this);
+            provider.RequestAnalysis(FilePath, Charset);
+        }
+
+        public void Dispose()
+        {
+            textBuffer.ChangedLowPriority -= this.OnBufferChange;
+            provider.RemoveIssueTracker(this);
+        }
+
+        private void OnBufferChange(object sender, TextContentChangedEventArgs e)
+        {
+            UpdateDirtySpans(e);
+
+            var newMarkers = TranslateMarkerSpans();
+
+            SnapToNewSnapshot(newMarkers);
+        }
+
+        private void UpdateDirtySpans(TextContentChangedEventArgs e)
+        {
+            currentSnapshot = e.After;
+
+            var newDirtySpans = dirtySpans.CloneAndTrackTo(e.After, SpanTrackingMode.EdgeInclusive);
+
+            foreach (var change in e.Changes)
+            {
+                newDirtySpans = NormalizedSnapshotSpanCollection.Union(newDirtySpans, new NormalizedSnapshotSpanCollection(e.After, change.NewSpan));
+            }
+
+            dirtySpans = newDirtySpans;
+        }
+
+        private IssuesSnapshot TranslateMarkerSpans()
+        {
+            var oldSnapshot = this.Factory.CurrentSnapshot;
+            var newMarkers = new List<IssueMarker>();
+
+            newMarkers.AddRange(oldSnapshot.IssueMarkers
+                .Select(marker => IssueMarker.CloneAndTranslateTo(marker, currentSnapshot))
+                .Where(clone => clone != null));
+
+            return new IssuesSnapshot(this.FilePath, oldSnapshot.VersionNumber + 1, newMarkers);
+        }
+
+        internal void UpdateIssues(IEnumerable<Issue> issues)
+        {
+            var oldSnapshot = this.Factory.CurrentSnapshot;
+            var newMarkers = new List<IssueMarker>();
+
+            foreach (Issue issue in issues)
+            {
+                int startPos = currentSnapshot.GetLineFromLineNumber(issue.StartLine - 1).Start.Position + issue.StartLineOffset;
+                var start = new SnapshotPoint(currentSnapshot, startPos);
+
+                int endPos = currentSnapshot.GetLineFromLineNumber(issue.EndLine - 1).Start.Position + issue.EndLineOffset;
+                var end = new SnapshotPoint(currentSnapshot, endPos);
+
+                newMarkers.Add(new IssueMarker(issue, new SnapshotSpan(start, end)));
+            }
```

Comment:
That code is fine, but I would extract the content of the loop to `CreateIssueMarker(Issue)` and then convert the loop to LINQ:

    var newMarkers = issues
        .Select(CreateIssueMarker)
        .ToList();

(that's equivalent to `issues.Select(issue => CreateIssueMarker(issue)).ToList()` but shorter)